### DISCUSSION TITLE
msys 编译问题

### DISCUSSION THREAD

**Discussion by hopenobug** [xmake-io/xmake#3489] [Q&A] [answered]
我在windows上使用tdm-gcc， xmake必须配置成msys平台才能正常编译
xmake-repo中很多package没有明确写msys， 就像下面这样的
on_install("linux", "macosx", "windows", function (package)

就导致msys无法编译

怎么在不替换tdm-gcc的情况下， 去编译这些呢？
如果一个个去修改package， 也不太现实
msys也是windows下的一种环境， 但是xmake把它从windows排除了，变成一个单独的平台

**Comment by waruqi** [comment, marked as answer]
需要提 pr 给特定包，加上 msys 平台，如果是 mingw 的。。得加上 mingw 

tdm-gcc 这种属于交叉编译，应该要加上 `on_install("cross"` ，然后切到 cross 平台编译，跟是否 msys 无关。。

```
xmake f -p cross --sdk=/xxx/tdm-gcc/xx
```

不同的平台，对包都有特殊处理。。windows 平台仅用于 msvc 工具链，ci上对 MD/MT/MDd/MTD/shared/static 有各种排列组合压测，也有预编译包优化。。还有特定 msvc 的patches

要是每个平台，还得随意切 toolchain 。。这个配置组合就要增长一个数量级。。我没这么多精力维护。而且包的配置脚本也会复杂一个数量级，也不能针对特定编译器做处理优化。。也会引入各种不确定性，也没法做 ci 测试，可靠性非常低

目前只有 cross 平台可以随意切工具链，xmake 会透传工具链给包内部的构建脚本，并且有专门一个交叉编译的 ci test 做一些基础测试。。

如果当前包没放开 cross，就自己提 pr 放开下，然后根据 ci 报错，自己处理下错误，通过后，我会 merge 。。
